src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -784,6 +812,266 @@ private void PopulateEncLogTableParameters()
+                // added in the previous generation
+                if (_customAttributeParentCounts[customAttributeTarget] == 0)
+                {
+                    // Expand our enc log table to accommodate
+                    encLogRows.Add(0);
+                    logIndex = encLogRows.Count - 1;
+                    // The custom attribute we're deleting came from a previous delta, so we don't have any data about it
+                    // so we just set its constructor to a MemberRef for row id 0
+                    AddDeletedCustomAttribute(customAttributeTarget.Kind, MetadataTokens.EntityHandle(TableIndex.MemberRef, 0), value: default);
+                }
+                else
+                {
+                    _customAttributeParentCounts[customAttributeTarget]--;
+                }
+
+                if (_customAttributeParentCounts[customAttributeTarget] == 0)
+                {
+                    // Finished with this parent
+                    index++;
+                }
+
+                // remainingEmittedAttributeParents only contains records that haven't been logged, but since we know it is in the
+                // right order, we don't need to know the original index we can just fill log rows in order, and skip any
+                // that were already filled in the previous phase.
+                Debug.Assert(logIndex < encLogRows.Count, "logIndex is invalid, possibly caused by a bug in this loop specifically. Is something incrementing it twice?");
+                while (encLogRows[logIndex] != 0)
+                {
+                    logIndex++;
+                }
+
+                Debug.Assert(rowId > 0, "Shouldn't ever emit a log row for custom attribute row 0");
+                encLogRows[logIndex] = rowId;
+            }
+        }
+
+        private void AddDeletedCustomAttribute(HandleKind parentKind, EntityHandle constructor, BlobHandle value)
+        {
+            // now emit a "delete" row with a parent that is for the 0 row of the same table as the existing one
+            var isValidKind = MetadataTokens.TryGetTableIndex(parentKind, out var tableIndex);
+
+            Debug.Assert(isValidKind, "Trying to delete a custom attribute for a parent kind that doesn't have a matching table index.");
+
+            if (isValidKind)
+            {
+                metadata.AddCustomAttribute(MetadataTokens.Handle(tableIndex, 0), constructor, value);
+            }
+        }
+
+        /// <summary>
+        /// For each custom attribute target that was emitted in previous generations, we need to keep a list of
+        /// which row numbers are used because whilst each generation is sorted by parent, we're looking across multiple
+        /// generations.
+        /// </summary>
+        private static Dictionary<EntityHandle, Queue<int>> CreateExistingAttributeMap(MetadataReader metadataReader, IReadOnlyDictionary<int, EntityHandle> customAttributesAdded, out int lastRowId)
```

Comment:
>out int lastRowId [](http://example.com/codeflow?start=180&length=17)

The parameter is a bit out of place.

Can't it be calculated in the caller simply by 
`Max(customAttributesAdded.Keys.Max(), metadataReader.GetTableRowCount(TableIndex.CustomAttribute))?`

Reply:
Yeah, I just had it here to avoid iterating the dictionary again.

Reply:
I guessed that's why, but I don't think that saves much and it conflates multiple functions in one.